pkg/config/kube/annotations.go

```diff
@@ -0,0 +1,138 @@
+		"Specifies one or more user volumes (as a JSON array) to be added to the istio-proxy sidecar.")
+
+	UserVolumeMountAnnotation = anno("sidecar.istio.io/userVolumeMount",
+		"Specifies one or more user volume mounts (as a JSON array) to be added to the istio-proxy sidecar.")
+
+	StatusPortAnnotation = anno("status.sidecar.istio.io/port",
+		"Specifies the HTTP status Port for the istio-proxy sidecar. If zero, the istio-proxy will not "+
+			"provide status.")
+
+	ReadinessInitialDelaySecondsAnnotation = anno("readiness.status.sidecar.istio.io/initialDelaySeconds",
+		"Specifies the initial delay (in seconds) for the istio-proxy readiness probe.")
+
+	ReadinessPeriodSecondsAnnotation = anno("readiness.status.sidecar.istio.io/periodSeconds",
+		"Specifies the period (in seconds) for the istio-proxy readiness probe.")
+
+	ReadinessFailureThresholdAnnotation = anno("readiness.status.sidecar.istio.io/failureThreshold",
+		"Specifies the failure threshold for the istio-proxy readiness probe.")
+
+	ReadinessApplicationPortsAnnotation = anno("readiness.status.sidecar.istio.io/applicationPorts",
+		"Specifies the list of ports exposed by the application container. Used by the istio-proxy "+
+			"readiness probe to determine that Envoy is configured and ready to receive traffic.")
+
+	TrafficIncludeOutboundIPRangesAnnotation = anno("traffic.sidecar.istio.io/includeOutboundIPRanges",
+		"A comma separated list of IP ranges in CIDR form to redirect to envoy (optional). The wildcard "+
+			"character '*' can be used to redirect all outbound traffic. An empty list will disable all outbound "+
+			"redirection")
+
+	TrafficExcludeOutboundIPRangesAnnotation = anno("traffic.sidecar.istio.io/excludeOutboundIPRanges",
+		"A comma separated list of IP ranges in CIDR form to be excluded from redirection. Only applies "+
+			"when all outbound traffic (i.e. '*') is being redirected")
+
+	TrafficIncludeInboundPortsAnnotation = anno("traffic.sidecar.istio.io/includeInboundPorts",
+		"A comma separated list of inbound ports for which traffic is to be redirected to Envoy. The"+
+			"wildcard character '*' can be used to configure redirection for all ports. An empty list will disable "+
+			"all inbound redirection")
+
+	TrafficExcludeInboundPortsAnnotation = anno("traffic.sidecar.istio.io/excludeInboundPorts",
+		"A comma separated list of inbound ports to be excluded from redirection to Envoy. Only applies "+
+			"when all inbound traffic (i.e. '*') is being redirected")
+
+	TrafficExcludeOutboundPortsAnnotation = anno("traffic.sidecar.istio.io/excludeOutboundPorts",
+		"A comma separated list of outbound ports to be excluded from redirection to Envoy")
+
+	TrafficKubeVirtInterfacesAnnotation = anno("traffic.sidecar.istio.io/kubevirtInterfaces",
+		"A comma separated list of virtual interfaces whose inbound traffic (from VM) will be treated "+
+			"as outbound")
+)
+
+var annoKeySet = make(map[string]struct{})
+
+func anno(name, desc string) annotations.Annotation {
```

Comment:
I'm not sure what the benefits are of consolidating all this stuff together, seems like breaking encapsulation. Kind of like taking all global variables from all components and putting them together in one file that everybody can use. What's the motivation?

We need to think about the impact of this model on how we document annotations. The current model is that we document annotations on the reference page of the component that declares the annotation. So if Pilot observes an annotation, the PIlot page documents this annotations.

With this new centralized approach to consolidating annotation declaration, all components will appear to use all annotations, which is not what we want.

Before this goes in then, we need to have a doc strategy.

Reply:
@geeknoid 

> I'm not sure what the benefits are of consolidating all this stuff together, seems like breaking encapsulation. Kind of like taking all global variables from all components and putting them together in one file that everybody can use. What's the motivation?

The goal here is to begin untangling some of the dependencies on pilot (e.g. galley, istioctl). For the most part our annotations share the same goal ... they drive the configuration of the sidecar. In that sense having them in a common place seems reasonable ... effectively, they're all just "Istio" annotations.

I'm open to location ... it just picked `pkg/config`, since the directory already existed and the annotations are included in the metadata of the k8s config resources.

> With this new centralized approach to consolidating annotation declaration, all components will appear to use all annotations, which is not what we want.

I think you're referring to the process of registering annotations. This PR only defines the annotations in a common location, but does not register them. They are still registered at the previous call sites.

Reply:
Any code importing this package will end up initializing all those variables. Initializing the variables ends up calling annotations.Annotation for every variable. This register  the existence of the annotation. When we generate reference documentation, the pkg/collateral package  is used to output per-binary docs. Part of what it does is look at all registered annotations and emits those in the reference docs for the binary. Given that all annotations are initialized, all annotations will show up in the docs for every binary using this package.

These  annotations are really an extension of our API.  Documenting the annotations in the reference page for each  binary is not the right thing to do, but it was the only reasonably simple option at the time. Ideally, I think annotations should be declared in the istio/api repo, alongside  our API protos. Then they could  show up as their own doc page on istio.io.

Just brainstorming here, but  we could have annotations.yaml in istio.io which defines all the supported annotations and provides descriptions. We could process this file to produce an .html page reference page and a .go file for use by our code. We could deprecate pkg/annotations completely (since that package exists strictly to enable pkg/collateral to generate docs  for annotations).

Reply:
OK, so Nathan  pointed out I was totally confused. This code is not  registering, it's merely declaring. Sorry about my confusion.

Reply:
@geeknoid I've just created https://github.com/istio/api/pull/979 to move these annotations to the API repository. PTAL